htmresearch/algorithms/extended_temporal_memory.py

```diff
@@ -21,538 +21,1134 @@
+    Amount by which permanences of synapses are decremented during learning.
+    @param learn (bool)
+    If true, grow and reinforce synapses.
+
+    @return cellsToAdd (list)
+    A list of predicted cells that will be added to active cells and winner
+    cells.
+    """
+    cellsToAdd = []
+
+    for cellData in groupby2(columnActiveBasalSegments, cellForSegment,
+                             columnMatchingBasalSegments, cellForSegment,
+                             columnActiveApicalSegments, cellForSegment,
+                             columnMatchingApicalSegments, cellForSegment):
+      (cell,
+       cellActiveBasalSegments, cellMatchingBasalSegments,
+       cellActiveApicalSegments, cellMatchingApicalSegments) = cellData
+
+      if predictiveScore(cellActiveBasalSegments,
+                         cellActiveApicalSegments) >= predictiveThreshold:
+        cellsToAdd.append(cell)
+
+        if learn:
+
+          # Do basal
+          internalBasalCandidates = (  # TODO: reformat this to be readable
+            prevWinnerCells if formInternalBasalConnections
+            else tuple())
+          ExtendedTemporalMemory.learnOnCell(
+            basalConnections,
+            random,
+            cell,
+            prevActiveCells,
+            internalBasalCandidates,
+            prevActiveExternalCellsBasal,
+            cellActiveBasalSegments,
+            cellMatchingBasalSegments,
+            maxNewSynapseCount,
+            initialPermanence,
+            permanenceIncrement,
+            permanenceDecrement)
+
+          # Do apical
+          internalApicalCandidates = tuple()
+          ExtendedTemporalMemory.learnOnCell(
+            apicalConnections, random,
+            cell,
+            cellActiveApicalSegments, cellMatchingApicalSegments,
+            prevActiveCells,
+            internalApicalCandidates, prevActiveExternalCellsApical,
+            numActivePotentialSynapsesForApicalSegment,
+            maxNewSynapseCount,
+            initialPermanence, permanenceIncrement, permanenceDecrement)
+
+    return cellsToAdd
+
+
+  @staticmethod
+  def learnOnCell(connections, random, cell, prevActiveCells,
+                  internalCandidates, externalCandidates, cellActiveSegments,
+                  cellMatchingSegments, maxNewSynapseCount, initialPermanence,
+                  permanenceIncrement, permanenceDecrement):
+    """
+
+    @param connections (Object)
+    Connections for the TM. Gets mutated.
+    @param random (Object)
+    Random number generator. Gets mutated.
+    """
+    # TODO: use lambda rather than itertools?
+    # bySegment = lambda segment: segment  #???
+    # for segmentData in groupBy2(cellActiveSegments, bySegment,
+    #                             cellMatchingSegments, bySegment):
+    #   # (segment,
+    #   #  x, y, z) = segmentData  # what is unpacked here?
+    #   segment = segmentData[0]
+    for segment in itertools.chain(cellActiveSegments, cellMatchingSegments):
```

Comment:
I doubt you'll ever want to concatenate that active segments and matching segments.